### PR TITLE
decomposedfs: shard nodes per space

### PR DESCRIPTION
**Pull request by butonic**
This changes the decomposedfs layout on disk and
- shards the nodes per space
- distributes the nodes over multiple directories to keep folder sizes manageable
- [ ] introduce subfolders for `spacetypes/{type}/{spaceid}`
- [ ] introduce supfolders for blobstorage -> separate PR
- [ ] migrate old `nodes` folder
- [ ] migrate old `spaces/(personal|share|project|metadata|...)` links? just recreate?

### CODE REVIEW

**Comment by micbar** [comment]
this needs a rebase.

**Comment by butonic** [comment]
> @butonic I suggest you to add constant with the width size rather than calling the numeric argument on each call. A width of 3 is plenty of room, the nodei-d is a uniform distribution or based on something like username/filename?

good points:

I initially wanted to make the depth and width configurable, but it has quite an effect on symlink parent segments and I will postpone that for now.

node ids should be a uuid v4 (random), so the distribution should be uniform.

That being said, personal spaces are created with the userid as the spaceid and the root node id to make debugging easier. We will have to assign an internal uuid or get one from the openid claims, as we need to attach roles to users anyway. But if eg. usernames (which are neither stable, reusable or persistend) are configured as uuids then the code will break for users with short names.

File: pkg/storage/utils/decomposedfs/recycle.go [hunk 1]
```diff
@@ -70,92 +71,94 @@ func (fs *Decomposedfs) ListRecycle(ctx context.Context, ref *provider.Reference
...
-	defer f.Close()
 
-	parentNode, err := os.Readlink(filepath.Join(trashRoot, key))
+	trashItemPath := filepath.Join(trashRootPath, relativePath)
+
+	f, err := os.Open(trashItemPath)
 	if err != nil {
-		log.Error().Err(err).Str("trashRoot", trashRoot).Msg("error reading trash link, skipping")
-		return nil, err
+		if os.IsNotExist(err) {
```
Comment (C0rby) [inline pkg/storage/utils/decomposedfs/recycle.go:112]:
```suggestion
		if errors.Is(err, fs.ErrNotExist) {
```
`os.IsNotExist` is deprecated. Don't forget to import `fs`.

File: pkg/storage/utils/decomposedfs/node/node.go [hunk 2]
```diff
@@ -135,6 +136,18 @@ func (n *Node) SetMetadata(key string, val string) (err error) {
 	return nil
 }
 
+// RemoveMetadata removes a given key
+func (n *Node) RemoveMetadata(key string) (err error) {
+	if err = xattrs.Remove(n.InternalPath(), key); err != nil {
+		if e, ok := err.(*xattr.Error); ok && (e.Err.Error() == "no data available" ||
```
[the remark this answers is not in the document]
Comment (butonic) [reply pkg/storage/utils/decomposedfs/node/node.go:142]: this was before we moved to the `xatts.IsAttrUnset` error check ... will update

File: pkg/storage/utils/decomposedfs/node/node.go [hunk 3]
```diff
@@ -238,12 +304,30 @@ func isNotDir(err error) bool {
 	return false
 }
 
+func readChildNodeFromLink(path string) (string, error) {
+	link, err := os.Readlink(path)
+	if err != nil {
+		return "", err
+	}
+	nodeID := strings.TrimLeft(link, "/.")
```
[the remark this answers is not in the document]
Comment (C0rby) [reply pkg/storage/utils/decomposedfs/node/node.go:308]: Ah right. :+1:

**Review by C0rby — APPROVED** [review]

**Pull request merged**